File: Buddy/buddy-ios/Sources/buddy/Models/Achievement.swift
import Foundation

enum Achievement: String, CaseIterable, Identifiable, Codable {
    case firstFeed
    case firstPlay
    case firstSleep
    case firstBath
    case fed10
    case fed50
    case played25
    case bath10
    case daysAlive3
    case daysAlive7
    case daysAlive30
    case maxMood
    case fullStats
    case coins100
    case coins500
    case minigame1
    case allMinigames
    case firstReincarnation

    var id: String { rawValue }

    var title: String {
        switch self {
        case .firstFeed:        "Primera comida"
        case .firstPlay:        "Primer juego"
        case .firstSleep:       "Primer sueño"
        case .firstBath:        "Primer baño"
        case .fed10:            "Cocinero · 10 comidas"
        case .fed50:            "Chef · 50 comidas"
        case .played25:         "Compañero · 25 juegos"
        case .bath10:           "Limpiador · 10 baños"
        case .daysAlive3:       "3 días vivo"
        case .daysAlive7:       "Una semana"
        case .daysAlive30:      "Un mes entero"
        case .maxMood:          "Máxima felicidad"
        case .fullStats:        "Cuidado perfecto"
        case .coins100:         "100 monedas"
        case .coins500:         "500 monedas"
        case .minigame1:        "Primer mini-juego"
        case .allMinigames:     "Todos los mini-juegos"
        case .firstReincarnation: "Primera reencarnación"
        }
    }

    var emoji: String {
        switch self {
        case .firstFeed:        "🍖"
        case .firstPlay:        "🎾"
        case .firstSleep:       "💤"
        case .firstBath:        "🛁"
        case .fed10:            "🥄"
        case .fed50:            "👨‍🍳"
        case .played25:         "🎮"
        case .bath10:           "🚿"
        case .daysAlive3:       "🌱"
        case .daysAlive7:       "🌿"
        case .daysAlive30:      "🌳"
        case .maxMood:          "😍"
        case .fullStats:        "⭐"
        case .coins100:         "💰"
        case .coins500:         "💎"
        case .minigame1:        "🎯"
        case .allMinigames:     "🏆"
        case .firstReincarnation: "🌟"
        }
    }

    var reward: Int {
        switch self {
        case .firstFeed, .firstPlay, .firstSleep, .firstBath, .minigame1: 5
        case .fed10, .played25, .bath10:                                  15
        case .daysAlive3, .coins100:                                      25
        case .daysAlive7, .coins500, .maxMood, .fullStats:                50
        case .fed50, .daysAlive30, .allMinigames, .firstReincarnation:    100
        }
    }
}

struct AchievementStore {
    private static let key = "buddy.achievements.v1"

    static func unlocked() -> Set<Achievement> {
        guard let data = UserDefaults.standard.data(forKey: key),
              let raws = try? JSONDecoder().decode([String].self, from: data) else { return [] }
        return Set(raws.compactMap { Achievement(rawValue: $0) })
    }

    @discardableResult
    static func unlock(_ a: Achievement) -> Bool {
        var set = unlocked()
        guard !set.contains(a) else { return false }
        set.insert(a)
        save(set)
        CoinWallet.add(a.reward)
        return true
    }

    private static func save(_ set: Set<Achievement>) {
        let raws = set.map(\.rawValue)
        if let data = try? JSONEncoder().encode(raws) {
            UserDefaults.standard.set(data, forKey: key)
        }
    }

    static func reset() {
        UserDefaults.standard.removeObject(forKey: key)
    }
}

File: Buddy/buddy-flutter/lib/models/achievement.dart
import 'package:shared_preferences/shared_preferences.dart';

import '../core/persistence/coin_wallet.dart';

enum Achievement {
  firstFeed,
  firstPlay,
  firstSleep,
  firstBath,
  fed10,
  fed50,
  played25,
  bath10,
  daysAlive3,
  daysAlive7,
  daysAlive30,
  maxMood,
  fullStats,
  coins100,
  coins500,
  minigame1,
  allMinigames,
  firstReincarnation;

  String get title => switch (this) {
        Achievement.firstFeed         => 'Primera comida',
        Achievement.firstPlay         => 'Primer juego',
        Achievement.firstSleep        => 'Primer sueño',
        Achievement.firstBath         => 'Primer baño',
        Achievement.fed10             => 'Cocinero · 10 comidas',
        Achievement.fed50             => 'Chef · 50 comidas',
        Achievement.played25          => 'Compañero · 25 juegos',
        Achievement.bath10            => 'Limpiador · 10 baños',
        Achievement.daysAlive3        => '3 días vivo',
        Achievement.daysAlive7        => 'Una semana',
        Achievement.daysAlive30       => 'Un mes entero',
        Achievement.maxMood           => 'Máxima felicidad',
        Achievement.fullStats         => 'Cuidado perfecto',
        Achievement.coins100          => '100 monedas',
        Achievement.coins500          => '500 monedas',
        Achievement.minigame1         => 'Primer mini-juego',
        Achievement.allMinigames      => 'Todos los mini-juegos',
        Achievement.firstReincarnation=> 'Primera reencarnación',
      };

  String get emoji => switch (this) {
        Achievement.firstFeed         => '🍖',
        Achievement.firstPlay         => '🎾',
        Achievement.firstSleep        => '💤',
        Achievement.firstBath         => '🛁',
        Achievement.fed10             => '🥄',
        Achievement.fed50             => '👨‍🍳',
        Achievement.played25          => '🎮',
        Achievement.bath10            => '🚿',
        Achievement.daysAlive3        => '🌱',
        Achievement.daysAlive7        => '🌿',
        Achievement.daysAlive30       => '🌳',
        Achievement.maxMood           => '😍',
        Achievement.fullStats         => '⭐',
        Achievement.coins100          => '💰',
        Achievement.coins500          => '💎',
        Achievement.minigame1         => '🎯',
        Achievement.allMinigames      => '🏆',
        Achievement.firstReincarnation=> '🌟',
      };

  int get reward => switch (this) {
        Achievement.firstFeed ||
        Achievement.firstPlay ||
        Achievement.firstSleep ||
        Achievement.firstBath ||
        Achievement.minigame1                                       => 5,
        Achievement.fed10 ||
        Achievement.played25 ||
        Achievement.bath10                                          => 15,
        Achievement.daysAlive3 ||
        Achievement.coins100                                        => 25,
        Achievement.daysAlive7 ||
        Achievement.coins500 ||
        Achievement.maxMood ||
        Achievement.fullStats                                       => 50,
        Achievement.fed50 ||
        Achievement.daysAlive30 ||
        Achievement.allMinigames ||
        Achievement.firstReincarnation                              => 100,
      };
}

class AchievementStore {
  static const _key = 'buddy.achievements.v1';

  static Future<Set<Achievement>> unlocked() async {
    final prefs = await SharedPreferences.getInstance();
    final raws = prefs.getStringList(_key) ?? const [];
    return raws
        .map((r) => Achievement.values.where((a) => a.name == r).firstOrNull)
        .whereType<Achievement>()
        .toSet();
  }

  /// Devuelve `true` si el achievement no estaba desbloqueado y se otorgó la recompensa.
  static Future<bool> unlock(Achievement a) async {
    final set = await unlocked();
    if (set.contains(a)) return false;
    set.add(a);
    final prefs = await SharedPreferences.getInstance();
    await prefs.setStringList(_key, set.map((e) => e.name).toList());
    await CoinWallet.add(a.reward);
    return true;
  }

  static Future<void> reset() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_key);
  }
}

extension<T> on Iterable<T> {
  T? get firstOrNull => isEmpty ? null : first;
}
